### PR TITLE
docs: move As documentation to doc.go

### PR DESCRIPTION
**Pull request by eliben**
This will make it easier for us to point to this documentation for more details from all the places where As is described

For #1274

### CODE REVIEW

File: doc.go [hunk 1]
```diff
@@ -35,5 +35,53 @@ increasing compile times or binary sizes, and avoiding any side effects from
...
+
+
+Escaping the abstraction
+
+It is not feasible or desirable for APIs like blob.Bucket to encompass the full
+functionality of every provider. Rather, we intend to provide a subset of the
+most commonly used functionality. There will be cases where a developer wants
+to access provider-specific functionality, which might consist of:
+
+1. Top-level APIs. For example, blob does not expose a Copy, but some provider
```
Comment (vangent) [inline doc.go]: WDYT about dropping this list, or collapsing it into a parenthetical? Seems like overkill for here.

File: doc.go [hunk 2]
```diff
@@ -35,5 +35,53 @@ increasing compile times or binary sizes, and avoiding any side effects from
...
+
+3. Errors. For example, S3 returns awserr.Error.
+
+4. Options. Different providers may support different options for functionality.
+
+As functions in the APIs provide the user a way to escape the Go CDK
+abstraction to access provider-specific types. They might be used as an interim
+solution until a feature request to the Go CDK is implemented. Or, the Go CDK
+may choose not to support specific features, and the use of As will be
+permanent. As an example, both S3 and GCS blobs have the concept of ACLs, but
```
Comment (vangent) [inline doc.go]: Same -- consider dropping the example?

File: doc.go [hunk 3]
```diff
@@ -35,5 +35,53 @@ increasing compile times or binary sizes, and avoiding any side effects from
...
+it might be difficult to represent them in a generic way (although, we have not
+tried).
+
+Using As implies that the resulting code is no longer portable; the
+provider-specific code will need to be ported in order to switch providers.
+Therefore, it should be avoided if possible. For example:
+
+	// The existing blob.Reader exposes some blob attributes, but not everything
```
Comment (vangent) [inline doc.go]: Drop the example and assert that each API should have specific examples for As?

File: doc.go (hunk 1 above)
Comment (eliben) [reply doc.go]: Good points

**Comment by eliben** [comment]
PTAL

File: doc.go [hunk 4]
```diff
@@ -42,46 +42,21 @@ Escaping the abstraction
...
-	var s3type s3.GetObjectOutput
-	if r.As(&s3type) {
-		... use s3type...
-	}
+Therefore, it should be avoided if possible.
 
-Each provider implementation documents what type(s) it supports for each of the
-As functions.
+Each provider implementation must document what type(s) it supports for each of
+the As functions.
```
Comment (vangent) [inline doc.go]: How about:

```
Each API will include examples demonstrating how to use its various As functions, and each provider implementation will document what types it supports for each.
```

?

Comment (eliben) [reply doc.go]: Done

**Pull request merged**